Sources/SwiftProtobuf/BinaryDecoder.swift

```diff
@@ -18,1477 +18,1505 @@
+
+      internal init(
+            forReadingFrom pointer: UnsafePointer<UInt8>, count: Int,
+            options: BinaryDecodingOptions, extensions: ExtensionMap? = nil
+      ) {
+            // Assuming baseAddress is not nil.
+            p = pointer
+            available = count
+            fieldStartP = p
+            self.extensions = extensions
+            self.options = options
+            recursionBudget = options.messageDepthLimit
+      }
+
+      internal init(
+            forReadingFrom pointer: UnsafePointer<UInt8>, count: Int,
+            parent: BinaryDecoder
+      ) {
+            self.init(
+                  forReadingFrom: pointer, count: count,
+                  options: parent.options, extensions: parent.extensions)
+            recursionBudget = parent.recursionBudget
+      }
+
+      private mutating func incrementRecursionDepth() throws {
+            recursionBudget -= 1
+            if recursionBudget < 0 {
+                  throw BinaryDecodingError.messageDepthLimit
             }
-        }
-
-        // Quit if end of input
-        if available == 0 {
-            return nil
-        }
-
-        // Get the next field number
-        fieldStartP = p
-        fieldEndP = nil
-        let start = p
-        let c0 = start[0]
-        if let wireFormat = WireFormat(rawValue: c0 & 7) {
-            fieldWireFormat = wireFormat
-        } else {
-            throw BinaryDecodingError.malformedProtobuf
-        }
-        if (c0 & 0x80) == 0 {
-            p += 1
-            available -= 1
-            fieldNumber = Int(c0) >> 3
-        } else {
-            fieldNumber = Int(c0 & 0x7f) >> 3
-            if available < 2 {
-                throw BinaryDecodingError.malformedProtobuf
+      }
+
+      private mutating func decrementRecursionDepth() {
+            recursionBudget += 1
+            // This should never happen, if it does, something is probably corrupting memory, and
+            // simply throwing doesn't make much sense.
+            if recursionBudget > options.messageDepthLimit {
+                  fatalError(
+                        "Somehow BinaryDecoding unwound more objects than it started"
+                  )
+            }
+      }
+
+      internal mutating func handleConflictingOneOf() throws {
+            /// Protobuf simply allows conflicting oneof values to overwrite
+      }
+
+      /// Return the next field number or nil if there are no more fields.
+      internal mutating func nextFieldNumber() throws -> Int? {
+            // Since this is called for every field, I've taken some pains
+            // to optimize it, including unrolling a tweaked version of
+            // the varint parser.
+            if fieldNumber > 0 {
+                  if let override = unknownOverride {
+                        assert(!options.discardUnknownFields)
+                        assert(
+                              fieldWireFormat != .startGroup && fieldWireFormat
```

Comment:
This should break before the `&&`

Reply:
Expressions involving multiple binary operators are somewhat tricky to deal with the way that SwiftSyntax exposes them. There's a single `ExprList` node that's just a flat list of alternating terms and operators, so the AST structure doesn't actually reflect the precedence of those operators (I assume this is because of support for custom operators and custom precedence).

We could solve this for the built-in operators by processing that list as a whole and grouping terms based on the precedence of those operators to get the desired wrapping, but it's not obvious what to do if we encounter a custom operator.